include/pstl_offload/internal/usm_memory_replacement_common.h

```diff
@@ -31,13 +32,96 @@ __is_power_of_two(std::size_t __number)
     return (__number != 0) && ((__number & __number - 1) == 0);
 }
 
+class __sycl_device_shared_ptr
+{
+    struct __shared_device
+    {
+        std::optional<sycl::device> _M_device;
+        // to keep reference to default context of the device as long as allocated memory objects exist
+        std::optional<sycl::context> _M_default_context;
+        std::atomic<std::size_t> _M_cnt;
+    };
+
+    __shared_device* _M_shared_device;
+
+  public:
+    template <typename _DeviceSelector>
+    __sycl_device_shared_ptr(const _DeviceSelector& __device_selector)
+        : _M_shared_device(new __shared_device{std::nullopt, std::nullopt, 1})
```

Comment:
To me, the version with explicit `__original_aligned_alloc` is less readable, because of placement new etc, and this is not a performance-critical path. `new` should work correctly in any place of code, so why not use it.

I added it as a separate commit to have a possibility to revert just in case.